File: bot/queue.py
"""
Persistent message queue backed by SQLite.

Messages are enqueued as 'pending', picked up by a worker,
moved to 'processing', then marked 'done' or 'failed'.
On startup, any 'processing' messages are reset to 'pending'
(they crashed mid-flight).
"""

import asyncio
import logging
import sqlite3
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from bot.config import PROJECT_ROOT

logger = logging.getLogger(__name__)

DB_PATH = PROJECT_ROOT / "data" / "queue.db"

# Statuses
PENDING = "pending"
PROCESSING = "processing"
DONE = "done"
FAILED = "failed"


@dataclass
class QueuedMessage:
    id: int
    sender_id: int
    chat_id: int
    msg_id: int
    text: str
    status: str
    created_at: float
    file_paths: str = ""  # JSON array of file paths, e.g. '["/tmp/sophia/img.png"]'


class MessageQueue:
    """SQLite-backed persistent message queue."""

    def __init__(self, db_path: Path = DB_PATH):
        self._db_path = db_path
        self._db_path.parent.mkdir(parents=True, exist_ok=True)
        self._conn = sqlite3.connect(str(db_path))
        self._conn.execute("PRAGMA journal_mode=WAL")
        self._conn.execute("PRAGMA synchronous=NORMAL")
        self._init_schema()
        self._event = asyncio.Event()

    def _init_schema(self):
        self._conn.execute("""
            CREATE TABLE IF NOT EXISTS messages (
                id          INTEGER PRIMARY KEY AUTOINCREMENT,
                sender_id   INTEGER NOT NULL,
                chat_id     INTEGER NOT NULL,
                msg_id      INTEGER NOT NULL,
                text        TEXT    NOT NULL,
                status      TEXT    NOT NULL DEFAULT 'pending',
                created_at  REAL    NOT NULL,
                updated_at  REAL    NOT NULL,
                file_paths  TEXT    NOT NULL DEFAULT ''
            )
        """)
        self._conn.execute("""
            CREATE INDEX IF NOT EXISTS idx_messages_status
            ON messages (status, created_at)
        """)
        self._conn.commit()
        # Migration: add file_paths column if missing
        cols = {r[1] for r in self._conn.execute("PRAGMA table_info(messages)").fetchall()}
        if "file_paths" not in cols:
            self._conn.execute("ALTER TABLE messages ADD COLUMN file_paths TEXT NOT NULL DEFAULT ''")
            self._conn.commit()

    def recover(self) -> int:
        """Reset 'processing' messages back to 'pending' after a crash.
        Returns number of recovered messages."""
        now = time.time()
        cur = self._conn.execute(
            "UPDATE messages SET status = ?, updated_at = ? WHERE status = ?",
            (PENDING, now, PROCESSING),
        )
        self._conn.commit()
        count = cur.rowcount
        if count:
            logger.info("Recovered %d messages stuck in 'processing'", count)
        return count

    def enqueue(self, sender_id: int, chat_id: int, msg_id: int, text: str, file_paths: str = "") -> tuple[int, bool]:
        """Add a message to the queue. Returns (queue_row_id, is_duplicate).
        Deduplicates: if the same sender+chat+text was enqueued within the last
        60 seconds and is still pending, returns the existing row instead.
        file_paths: JSON array string of local file paths, e.g. '["/tmp/sophia/img.png"]'
        """
        now = time.time()

        # Dedup: check for identical pending message from same sender+chat within 60s
        cutoff = now - 60
        existing = self._conn.execute(
            "SELECT id FROM messages WHERE sender_id = ? AND chat_id = ? AND text = ? "
            "AND status = ? AND created_at > ? ORDER BY created_at DESC LIMIT 1",
            (sender_id, chat_id, text, PENDING, cutoff),
        ).fetchone()
        if existing:
            logger.info("Dedup: skipping duplicate message from %d (existing queue id %d)", sender_id, existing[0])
            return existing[0], True

        cur = self._conn.execute(
            "INSERT INTO messages (sender_id, chat_id, msg_id, text, status, created_at, updated_at, file_paths) "
            "VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
            (sender_id, chat_id, msg_id, text, PENDING, now, now, file_paths),
        )
        self._conn.commit()
        row_id = cur.lastrowid
        logger.debug("Enqueued message %d from %d (files: %s)", row_id, sender_id, bool(file_paths))
        self._event.set()
        return row_id, False

    def take_next(self) -> Optional[QueuedMessage]:
        """Atomically take the next pending message (FIFO).
        Returns None if queue is empty."""
        now = time.time()
        row = self._conn.execute(
            "SELECT id, sender_id, chat_id, msg_id, text, status, created_at, file_paths "
            "FROM messages WHERE status = ? ORDER BY created_at ASC LIMIT 1",
            (PENDING,),
        ).fetchone()
        if not row:
            return None
        msg = QueuedMessage(*row)
        self._conn.execute(
            "UPDATE messages SET status = ?, updated_at = ? WHERE id = ?",
            (PROCESSING, now, msg.id),
        )
        self._conn.commit()
        return msg

    def peek_next(self) -> Optional[QueuedMessage]:
        """Peek at the next pending message without changing its status."""
        row = self._conn.execute(
            "SELECT id, sender_id, chat_id, msg_id, text, status, created_at, file_paths "
            "FROM messages WHERE status = ? ORDER BY created_at ASC LIMIT 1",
            (PENDING,),
        ).fetchone()
        if not row:
            return None
        return QueuedMessage(*row)

    def take_batch(self, sender_id: int, chat_id: int) -> list[QueuedMessage]:
        """Atomically take ALL pending messages from a given sender+chat.
        Returns them in FIFO order."""
        now = time.time()
        rows = self._conn.execute(
            "SELECT id, sender_id, chat_id, msg_id, text, status, created_at, file_paths "
            "FROM messages WHERE status = ? AND sender_id = ? AND chat_id = ? "
            "ORDER BY created_at ASC",
            (PENDING, sender_id, chat_id),
        ).fetchall()
        if not rows:
            return []
        msgs = [QueuedMessage(*r) for r in rows]
        ids = [m.id for m in msgs]
        placeholders = ",".join("?" * len(ids))
        self._conn.execute(
            f"UPDATE messages SET status = ?, updated_at = ? WHERE id IN ({placeholders})",
            [PROCESSING, now] + ids,
        )
        self._conn.commit()
        return msgs

    def latest_pending_time(self, sender_id: int, chat_id: int) -> Optional[float]:
        """Return created_at of the most recent pending message from sender+chat."""
        row = self._conn.execute(
            "SELECT created_at FROM messages WHERE status = ? AND sender_id = ? AND chat_id = ? "
            "ORDER BY created_at DESC LIMIT 1",
            (PENDING, sender_id, chat_id),
        ).fetchone()
        return row[0] if row else None

    def mark_done(self, msg_id: int):
        self._conn.execute(
            "UPDATE messages SET status = ?, updated_at = ? WHERE id = ?",
            (DONE, time.time(), msg_id),
        )
        self._conn.commit()

    def mark_failed(self, msg_id: int):
        self._conn.execute(
            "UPDATE messages SET status = ?, updated_at = ? WHERE id = ?",
            (FAILED, time.time(), msg_id),
        )
        self._conn.commit()

    def pending_count(self) -> int:
        row = self._conn.execute(
            "SELECT COUNT(*) FROM messages WHERE status IN (?, ?)",
            (PENDING, PROCESSING),
        ).fetchone()
        return row[0]

    def cleanup(self, older_than_hours: int = 24):
        """Remove completed messages older than N hours."""
        cutoff = time.time() - older_than_hours * 3600
        cur = self._conn.execute(
            "DELETE FROM messages WHERE status = ? AND updated_at < ?",
            (DONE, cutoff),
        )
        self._conn.commit()
        if cur.rowcount:
            logger.info("Cleaned up %d old messages from queue", cur.rowcount)

    async def wait_for_messages(self, timeout: float = 2.0):
        """Wait until a new message is enqueued or timeout."""
        self._event.clear()
        try:
            await asyncio.wait_for(self._event.wait(), timeout=timeout)
        except asyncio.TimeoutError:
            pass

    def close(self):
        self._conn.close()

File: src/queue.rs
#![allow(dead_code)]

use std::path::Path;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

use anyhow::{Context, Result};
use rusqlite::Connection;
use tracing::info;

fn now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs_f64()
}

#[derive(Debug, Clone)]
pub struct QueuedMessage {
    pub id: i64,
    pub sender_id: i64,
    pub chat_id: i64,
    pub msg_id: i32,
    pub text: String,
    pub status: String,
    pub created_at: f64,
    pub file_paths: String,
}

#[derive(Clone)]
pub struct MessageQueue {
    conn: Arc<Mutex<Connection>>,
}

impl MessageQueue {
    pub fn new(db_path: &Path) -> Result<Self> {
        if let Some(parent) = db_path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let conn = Connection::open(db_path).context("Failed to open queue database")?;
        conn.execute_batch("PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL;")?;
        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS messages (
                id          INTEGER PRIMARY KEY AUTOINCREMENT,
                sender_id   INTEGER NOT NULL,
                chat_id     INTEGER NOT NULL,
                msg_id      INTEGER NOT NULL,
                text        TEXT    NOT NULL,
                status      TEXT    NOT NULL DEFAULT 'pending',
                created_at  REAL    NOT NULL,
                updated_at  REAL    NOT NULL,
                file_paths  TEXT    NOT NULL DEFAULT ''
            );
            CREATE INDEX IF NOT EXISTS idx_messages_status
            ON messages (status, created_at);
            CREATE INDEX IF NOT EXISTS idx_messages_chat_msg
            ON messages (chat_id, msg_id);",
        )?;

        // Migration: add file_paths if missing
        let has_file_paths = conn
            .prepare("PRAGMA table_info(messages)")?
            .query_map([], |row| row.get::<_, String>(1))?
            .filter_map(|r| r.ok())
            .any(|name| name == "file_paths");
        if !has_file_paths {
            conn.execute(
                "ALTER TABLE messages ADD COLUMN file_paths TEXT NOT NULL DEFAULT ''",
                [],
            )?;
        }

        Ok(Self {
            conn: Arc::new(Mutex::new(conn)),
        })
    }

    pub fn recover(&self) -> Result<usize> {
        let conn = self.conn.lock().unwrap();
        let t = now();
        let count = conn.execute(
            "UPDATE messages SET status = 'pending', updated_at = ?1 WHERE status = 'processing'",
            [t],
        )?;
        if count > 0 {
            info!("Recovered {} messages stuck in 'processing'", count);
        }
        Ok(count)
    }

    pub fn enqueue(
        &self,
        sender_id: i64,
        chat_id: i64,
        msg_id: i32,
        text: &str,
        file_paths: &str,
    ) -> Result<(i64, bool)> {
        let conn = self.conn.lock().unwrap();
        let t = now();
        let cutoff = t - 60.0;

        // Dedup by msg_id: skip if this Telegram message exists in ANY status
        // (pending, processing, or done) — prevents re-processing after restart + catch_up
        let already_handled: bool = conn
            .query_row(
                "SELECT COUNT(*) FROM messages WHERE chat_id = ?1 AND msg_id = ?2",
                rusqlite::params![chat_id, msg_id],
                |row| row.get::<_, i64>(0),
            )
            .unwrap_or(0)
            > 0;

        if already_handled {
            info!("Skipping duplicate msg_id={} (already handled)", msg_id);
            return Ok((0, true));
        }

        // Dedup by text content (for batching window)
        // Check both 'pending' AND 'processing' — prevents re-enqueue while Claude is still thinking
        let existing: Option<i64> = conn
            .query_row(
                "SELECT id FROM messages WHERE sender_id = ?1 AND chat_id = ?2 AND text = ?3 \
                 AND status IN ('pending', 'processing') AND created_at > ?4 ORDER BY created_at DESC LIMIT 1",
                rusqlite::params![sender_id, chat_id, text, cutoff],
                |row| row.get(0),
            )
            .ok();

        if let Some(id) = existing {
            return Ok((id, true));
        }

        conn.execute(
            "INSERT INTO messages (sender_id, chat_id, msg_id, text, status, created_at, updated_at, file_paths) \
             VALUES (?1, ?2, ?3, ?4, 'pending', ?5, ?5, ?6)",
            rusqlite::params![sender_id, chat_id, msg_id, text, t, file_paths],
        )?;
        let id = conn.last_insert_rowid();
        Ok((id, false))
    }

    pub fn take_batch(&self, sender_id: i64, chat_id: i64) -> Result<Vec<QueuedMessage>> {
        let conn = self.conn.lock().unwrap();
        let t = now();

        let mut stmt = conn.prepare(
            "SELECT id, sender_id, chat_id, msg_id, text, status, created_at, file_paths \
             FROM messages WHERE status = 'pending' AND sender_id = ?1 AND chat_id = ?2 \
             ORDER BY created_at ASC",
        )?;
        let msgs: Vec<QueuedMessage> = stmt
            .query_map(rusqlite::params![sender_id, chat_id], |row| {
                Ok(QueuedMessage {
                    id: row.get(0)?,
                    sender_id: row.get(1)?,
                    chat_id: row.get(2)?,
                    msg_id: row.get(3)?,
                    text: row.get(4)?,
                    status: row.get(5)?,
                    created_at: row.get(6)?,
                    file_paths: row.get(7)?,
                })
            })?
            .filter_map(|r| r.ok())
            .collect();

        if !msgs.is_empty() {
            let ids: Vec<String> = msgs.iter().map(|m| m.id.to_string()).collect();
            let placeholders = ids.join(",");
            conn.execute(
                &format!(
                    "UPDATE messages SET status = 'processing', updated_at = {} WHERE id IN ({})",
                    t, placeholders
                ),
                [],
            )?;
        }

        Ok(msgs)
    }

    pub fn mark_done(&self, id: i64) -> Result<()> {
        let conn = self.conn.lock().unwrap();
        conn.execute(
            "UPDATE messages SET status = 'done', updated_at = ?1 WHERE id = ?2",
            rusqlite::params![now(), id],
        )?;
        Ok(())
    }

    pub fn mark_failed(&self, id: i64) -> Result<()> {
        let conn = self.conn.lock().unwrap();
        conn.execute(
            "UPDATE messages SET status = 'failed', updated_at = ?1 WHERE id = ?2",
            rusqlite::params![now(), id],
        )?;
        Ok(())
    }

    pub fn cleanup(&self, older_than_hours: u64) -> Result<usize> {
        let conn = self.conn.lock().unwrap();
        let cutoff = now() - (older_than_hours as f64 * 3600.0);
        let count = conn.execute(
            "DELETE FROM messages WHERE status = 'done' AND updated_at < ?1",
            [cutoff],
        )?;
        if count > 0 {
            info!("Cleaned up {} old messages from queue", count);
        }
        Ok(count)
    }
}
